File: examples/experimental/agent_builder/customer_care_style/toolkits/banking_mcp_server/py_server/src/personal_banking_service.py
"""
Personal Banking Service - Mock Implementation

Contains mock business logic for personal banking operations.
"""

from dataclasses import dataclass
from datetime import datetime
import random
import string


@dataclass
class Account:
    """Bank account information."""

    account_id: str
    account_name: str
    account_type: str
    can_transfer_from: bool
    can_transfer_to: bool
    is_locked: bool
    account_holder: str | None = None


@dataclass
class Transaction:
    """Transaction record."""

    date: str
    description: str
    amount: float
    balance: float


@dataclass
class TransferResult:
    """Result of a money transfer."""

    customer_id: str
    from_account_id: str
    to_account_id: str
    amount: float
    transfer_id: str
    status: str
    timestamp: str
    scheduled_date: str | None = None
    memo: str | None = None


class PersonalBankingService:
    """Mock Personal Banking Service."""

    @staticmethod
    def get_accounts(customer_id: str) -> list[Account]:
        """
        Get list of accounts for a customer.
        Includes owned accounts and registered external accounts for transfers.
        """
        # Mock account data - in a real system, this would query a database
        return [
            Account(
                account_id="acc_checking_001",
                account_name="Primary Checking",
                account_type="checking",
                can_transfer_from=True,
                can_transfer_to=True,
                is_locked=True,  # This account is locked - requires customer service
            ),
            Account(
                account_id="acc_savings_001",
                account_name="High Yield Savings",
                account_type="savings",
                can_transfer_from=True,
                can_transfer_to=True,
                is_locked=False,
            ),
            Account(
                account_id="acc_checking_002",
                account_name="Business Checking",
                account_type="checking",
                can_transfer_from=True,
                can_transfer_to=True,
                is_locked=False,
            ),
            Account(
                account_id="acc_investment_001",
                account_name="Investment Account",
                account_type="investment",
                can_transfer_from=False,  # Cannot transfer from investment accounts
                can_transfer_to=True,
                is_locked=False,
            ),
            Account(
                account_id="acc_external_son_001",
                account_name="Son's College Fund",
                account_type="external_savings",
                can_transfer_from=False,  # Cannot transfer from son's account
                can_transfer_to=True,
                is_locked=False,
                account_holder="Michael Johnson (Son)",
            ),
            Account(
                account_id="acc_investment_002",
                account_name="Retirement IRA",
                account_type="retirement",
                can_transfer_from=False,  # Cannot transfer from retirement accounts
                can_transfer_to=True,
                is_locked=False,
            ),
        ]

    @staticmethod
    def get_account_balance(
        customer_id: str, account_type: str
    ) -> dict:
        """Get account balance by account type."""
        # Mock balance data
        balance = 5432.18 if account_type == "checking" else 12750.50
        return {
            "customer_id": customer_id,
            "account_type": account_type,
            "balance": balance,
            "currency": "USD",
            "as_of_date": datetime.now().isoformat(),
        }

    @staticmethod
    def get_account_statement(
        customer_id: str, account_type: str, days: int = 30
    ) -> dict:
        """Get account statement with transactions."""
        end_date = datetime.now()
        start_date = datetime.now()
        # Simplified date calculation for mock data

        # Mock transaction data
        transactions = [
            Transaction(
                date="2025-12-18",
                description="Direct Deposit - Salary",
                amount=3500.00,
                balance=5432.18,
            ),
            Transaction(
                date="2025-12-15",
                description="Online Purchase - Amazon",
                amount=-87.43,
                balance=1932.18,
            ),
            Transaction(
                date="2025-12-12",
                description="ATM Withdrawal",
                amount=-100.00,
                balance=2019.61,
            ),
            Transaction(
                date="2025-12-10",
                description="Transfer to Savings",
                amount=-500.00,
                balance=2119.61,
            ),
        ]

        return {
            "customer_id": customer_id,
            "account_type": account_type,
            "start_date": start_date.strftime("%Y-%m-%d"),
            "end_date": end_date.strftime("%Y-%m-%d"),
            "transactions": [
                {
                    "date": t.date,
                    "description": t.description,
                    "amount": t.amount,
                    "balance": t.balance,
                }
                for t in transactions
            ],
        }

    @staticmethod
    def transfer_money(
        customer_id: str,
        from_account_id: str,
        to_account_id: str,
        amount: float,
        scheduled_date: str | None = None,
        memo: str | None = None,
    ) -> TransferResult:
        """Transfer money between accounts."""
        # Validate that from and to accounts are different
        if from_account_id == to_account_id:
            raise ValueError("Cannot transfer to the same account")

        # Generate a mock transfer ID
        random_suffix = "".join(random.choices(string.ascii_lowercase + string.digits, k=9))
        transfer_id = f"TXN-{int(datetime.now().timestamp() * 1000)}-{random_suffix}"
        timestamp = datetime.now().isoformat()
        status = "scheduled" if scheduled_date else "completed"

        return TransferResult(
            customer_id=customer_id,
            from_account_id=from_account_id,
            to_account_id=to_account_id,
            amount=amount,
            scheduled_date=scheduled_date,
            memo=memo,
            transfer_id=transfer_id,
            status=status,
            timestamp=timestamp,
        )

    @staticmethod
    def validate_transfer(
        customer_id: str,
        from_account_id: str,
        to_account_id: str,
        amount: float,
    ) -> dict:
        """
        Validate a transfer request.
        Returns account details if valid, raises error if invalid.
        """
        accounts = PersonalBankingService.get_accounts(customer_id)
        from_account = next((acc for acc in accounts if acc.account_id == from_account_id), None)
        to_account = next((acc for acc in accounts if acc.account_id == to_account_id), None)

        if not from_account:
            raise ValueError(f"Source account {from_account_id} not found")

        if not to_account:
            raise ValueError(f"Destination account {to_account_id} not found")

        if not from_account.can_transfer_from:
            raise ValueError(
                f"Cannot transfer from {from_account.account_name} - transfers not allowed"
            )

        if not to_account.can_transfer_to:
            raise ValueError(
                f"Cannot transfer to {to_account.account_name} - transfers not allowed"
            )

        if amount <= 0:
            raise ValueError("Transfer amount must be greater than zero")

        return {"from_account": from_account, "to_account": to_account}

File: examples/experimental/agent_builder/customer_care_style/toolkits/banking_mcp_server/ts_server/src/personalBankingService.ts
/**
 * Personal Banking Service - Mock Implementation
 *
 * Contains mock business logic for personal banking operations
 */

export interface Account {
  accountId: string;
  accountName: string;
  accountType: string;
  canTransferFrom: boolean;
  canTransferTo: boolean;
  isLocked: boolean; // Indicates if the account is locked
  accountHolder?: string; // Optional field for accounts held by others (e.g., son's account)
}

export interface Transaction {
  date: string;
  description: string;
  amount: number;
  balance: number;
}

export interface TransferResult {
  customerId: string;
  fromAccountId: string;
  toAccountId: string;
  amount: number;
  scheduledDate?: string;
  memo?: string;
  transferId: string;
  status: string;
  timestamp: string;
}

export interface PreparedTransfer {
  transactionId: string;
  fromAccountId: string;
  fromAccountName: string;
  toAccountId: string;
  toAccountName: string;
  amount: number;
  scheduledDate?: string;
  memo?: string;
}

/**
 * Mock Personal Banking Service
 */
export class PersonalBankingService {
  /**
   * Get list of accounts for a customer
   * Includes owned accounts and registered external accounts for transfers
   */
  static getAccounts(customerId: string): Account[] {
    // Mock account data - in a real system, this would query a database
    return [
      {
        accountId: 'acc_checking_001',
        accountName: 'Primary Checking',
        accountType: 'checking',
        canTransferFrom: true,
        canTransferTo: true,
        isLocked: true, // This account is locked - requires customer service
      },
      {
        accountId: 'acc_savings_001',
        accountName: 'High Yield Savings',
        accountType: 'savings',
        canTransferFrom: true,
        canTransferTo: true,
        isLocked: false,
      },
      {
        accountId: 'acc_checking_002',
        accountName: 'Business Checking',
        accountType: 'checking',
        canTransferFrom: true,
        canTransferTo: true,
        isLocked: false,
      },
      {
        accountId: 'acc_investment_001',
        accountName: 'Investment Account',
        accountType: 'investment',
        canTransferFrom: false, // Cannot transfer from investment accounts
        canTransferTo: true,
        isLocked: false,
      },
      {
        accountId: 'acc_external_son_001',
        accountName: "Son's College Fund",
        accountType: 'external_savings',
        canTransferFrom: false, // Cannot transfer from son's account
        canTransferTo: true,
        isLocked: false,
        accountHolder: 'Michael Johnson (Son)',
      },
      {
        accountId: 'acc_investment_002',
        accountName: 'Retirement IRA',
        accountType: 'retirement',
        canTransferFrom: false, // Cannot transfer from retirement accounts
        canTransferTo: true,
        isLocked: false,
      },
    ];
  }

  /**
   * Get account balance by account type
   */
  static getAccountBalance(
    customerId: string,
    accountType: string,
  ): {
    customerId: string;
    accountType: string;
    balance: number;
    currency: string;
    asOfDate: string;
  } {
    // Mock balance data
    const balance = accountType === 'checking' ? 5432.18 : 12750.5;
    return {
      customerId,
      accountType,
      balance,
      currency: 'USD',
      asOfDate: new Date().toISOString(),
    };
  }

  /**
   * Get account statement with transactions
   */
  static getAccountStatement(
    customerId: string,
    accountType: string,
    days: number = 30,
  ): {
    customerId: string;
    accountType: string;
    startDate: string;
    endDate: string;
    transactions: Transaction[];
  } {
    const endDate = new Date();
    const startDate = new Date();
    startDate.setDate(startDate.getDate() - days);

    // Mock transaction data
    const transactions: Transaction[] = [
      {
        date: '2025-12-18',
        description: 'Direct Deposit - Salary',
        amount: 3500.0,
        balance: 5432.18,
      },
      {
        date: '2025-12-15',
        description: 'Online Purchase - Amazon',
        amount: -87.43,
        balance: 1932.18,
      },
      {
        date: '2025-12-12',
        description: 'ATM Withdrawal',
        amount: -100.0,
        balance: 2019.61,
      },
      {
        date: '2025-12-10',
        description: 'Transfer to Savings',
        amount: -500.0,
        balance: 2119.61,
      },
    ];

    return {
      customerId,
      accountType,
      startDate: startDate.toISOString().split('T')[0],
      endDate: endDate.toISOString().split('T')[0],
      transactions,
    };
  }

  /**
   * Transfer money between accounts
   */
  static transferMoney(
    customerId: string,
    fromAccountId: string,
    toAccountId: string,
    amount: number,
    scheduledDate?: string,
    memo?: string,
  ): TransferResult {
    // Validate that from and to accounts are different
    if (fromAccountId === toAccountId) {
      throw new Error('Cannot transfer to the same account');
    }

    // Generate a mock transfer ID
    const transferId = `TXN-${Date.now()}-${Math.random().toString(36).substr(2, 9)}`;
    const timestamp = new Date().toISOString();
    const status =
      scheduledDate !== undefined && scheduledDate !== ''
        ? 'scheduled'
        : 'completed';

    return {
      customerId,
      fromAccountId,
      toAccountId,
      amount,
      scheduledDate,
      memo,
      transferId,
      status,
      timestamp,
    };
  }

  /**
   * Validate a transfer request
   * Returns account details if valid, throws error if invalid
   */
  static validateTransfer(
    customerId: string,
    fromAccountId: string,
    toAccountId: string,
    amount: number,
  ): {
    fromAccount: Account;
    toAccount: Account;
  } {
    const accounts = this.getAccounts(customerId);
    const fromAccount = accounts.find(acc => acc.accountId === fromAccountId);
    const toAccount = accounts.find(acc => acc.accountId === toAccountId);

    if (!fromAccount) {
      throw new Error(`Source account ${fromAccountId} not found`);
    }

    if (!toAccount) {
      throw new Error(`Destination account ${toAccountId} not found`);
    }

    if (!fromAccount.canTransferFrom) {
      throw new Error(
        `Cannot transfer from ${fromAccount.accountName} - transfers not allowed`,
      );
    }

    if (!toAccount.canTransferTo) {
      throw new Error(
        `Cannot transfer to ${toAccount.accountName} - transfers not allowed`,
      );
    }

    if (amount <= 0) {
      throw new Error('Transfer amount must be greater than zero');
    }

    return { fromAccount, toAccount };
  }

  /**
   * Get mock transaction for confirmation/cancellation
   * In a real implementation, this would look up the transaction from storage
   */
  static getMockTransaction(customerId: string, transactionId: string) {
    const accounts = this.getAccounts(customerId);
    return {
      customerId,
      fromAccountId: accounts[0].accountId,
      fromAccountName: accounts[0].accountName,
      toAccountId: accounts[1].accountId,
      toAccountName: accounts[1].accountName,
      amount: 500.0,
      scheduledDate: new Date().toISOString().split('T')[0],
    };
  }
}
